x/capability/capability_test.go

```diff
@@ -0,0 +1,86 @@
+package capability_test
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/suite"
+	abci "github.com/tendermint/tendermint/abci/types"
+	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/cosmos/cosmos-sdk/simapp"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/module"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	"github.com/cosmos/cosmos-sdk/x/capability"
+	"github.com/cosmos/cosmos-sdk/x/capability/keeper"
+	"github.com/cosmos/cosmos-sdk/x/capability/types"
+)
+
+type CapabilityTestSuite struct {
+	suite.Suite
+
+	cdc    codec.Marshaler
+	ctx    sdk.Context
+	app    *simapp.SimApp
+	keeper *keeper.Keeper
+	module module.AppModule
+}
+
+func (suite *CapabilityTestSuite) SetupTest() {
+	checkTx := false
+	app := simapp.Setup(checkTx)
+	cdc := app.AppCodec()
+
+	// create new keeper so we can define custom scoping before init and seal
+	keeper := keeper.NewKeeper(cdc, app.GetKey(types.StoreKey), app.GetMemKey(types.MemStoreKey))
+
+	suite.app = app
+	suite.ctx = app.BaseApp.NewContext(checkTx, tmproto.Header{Height: 1})
+	suite.keeper = keeper
+	suite.cdc = cdc
+	suite.module = capability.NewAppModule(cdc, *keeper)
+}
+
+// The following test case mocks a specific bug discovered in https://github.com/cosmos/cosmos-sdk/issues/9800
+// and ensures that the current code successfully fixes the issue.
+func (suite *CapabilityTestSuite) TestInitializeMemStore() {
+	sk1 := suite.keeper.ScopeToModule(banktypes.ModuleName)
+
+	cap1, err := sk1.NewCapability(suite.ctx, "transfer")
+	suite.Require().NoError(err)
+	suite.Require().NotNil(cap1)
+
+	// mock statesync by creating new keeper that shares persistent state but loses in-memory map
+	newKeeper := keeper.NewKeeper(suite.cdc, suite.app.GetKey(types.StoreKey), suite.app.GetMemKey("testing"))
+	newSk1 := newKeeper.ScopeToModule(banktypes.ModuleName)
+
+	// Mock App startup
+	ctx := suite.app.BaseApp.NewUncachedContext(false, tmproto.Header{})
+	newKeeper.InitializeAndSeal(ctx)
+
+	// Mock app beginblock and ensure that no gas has been consumed
+	ctx = suite.app.BaseApp.NewContext(false, tmproto.Header{}).WithBlockGasMeter(sdk.NewGasMeter(50))
+	prevGas := ctx.BlockGasMeter().GasConsumed()
+	restartedModule := capability.NewAppModule(suite.cdc, *newKeeper)
+	restartedModule.BeginBlock(ctx, abci.RequestBeginBlock{})
+	gasUsed := ctx.BlockGasMeter().GasConsumed()
+
+	suite.Require().Equal(prevGas, gasUsed, "beginblocker consumed gas during execution")
+
+	// Mock the first transaction getting capability and subsequently failing
+	// by using a cached context and discarding all cached writes.
+	cacheCtx, _ := ctx.CacheContext()
+	_, ok := newSk1.GetCapability(cacheCtx, "transfer")
+	suite.Require().True((ok))
```

Comment:
shouldn't this be false? also what does `(ok)` do?

I think this needs to be moved up before the begin blocker section

Reply:
`(ok)` was a typo sorry.

No this should be after the beginblocker and it should pass. This is mocking a situation where the capability check on first tx after restart passes but rest of tx fails (hence discarding the cached ctx), and then calling a second tx.

Note the test case is no longer even testing anything relevant now that this fix is in, but it would reliably fail on v0.42.7. I've left it here to act as a canary in case, any future changes reintroduce the bug